core/domain/user_jobs_one_off.py

```diff
@@ -470,3 +470,120 @@ def map(user_settings_model):
     def reduce(key, values):
         """Implements the reduce function for this job."""
         yield (key, len(values))
+
+
+class CleanupUserSubscribersModelOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+    """Job that cleans up UserSubscribersModel by removing user id if it is
+    present in subscriber ids.
+    """
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [user_models.UserSubscribersModel]
+
+    @staticmethod
+    def map(item):
+        if not item.deleted and item.id in item.subscriber_ids:
+            item.subscriber_ids.remove(item.id)
+            item.put()
+            yield ('Cleaned up User Subscribers', 1)
+
+    @staticmethod
+    def reduce(key, values):
+        yield (key, len(values))
+
+
+class CleanupCollectionProgressModelOneOffJob(
+        jobs.BaseMapReduceOneOffJobManager):
+    """Job that cleans up CollectionProgressModel by:
+    1. Removing exploration ids which are not a part of the collection.
+    2. Creating CompletedActivitiesModel for completed explorations if it is
+    missing.
+    """
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [user_models.CollectionProgressModel]
+
+    @staticmethod
+    def map(item):
+        if not item.deleted:
+            col_model = collection_models.CollectionModel.get_by_id(
+                item.collection_id)
+
+            collection_node_ids = [
+                node['exploration_id'] for node in (
+                    col_model.collection_contents['nodes'])]
+            exp_ids_to_remove = [
+                exp_id
+                for exp_id in item.completed_explorations if exp_id not in (
+                    collection_node_ids)]
+
+            if exp_ids_to_remove:
+                item.completed_explorations = [
+                    exp_id for exp_id in item.completed_explorations
+                    if exp_id not in exp_ids_to_remove]
+                item.put()
+                yield (
+                    'Invalid Exploartion IDs cleaned from '
+                    'CollectionProgressModel', 1)
+
+            completed_activities_model = (
+                user_models.CompletedActivitiesModel.get_by_id(item.user_id))
+            if completed_activities_model is None or (
+                    completed_activities_model.deleted):
+                completed_activities_model = (
+                    user_models.CompletedActivitiesModel(id=item.user_id))
+                completed_activities_model.exploration_ids = (
```

Comment:
This and the following lines should go outside the "if", right?

Also if the model is deleted ... are you planning to undelete it? I'm a bit confused about that...

Maybe this needs a bit more thought.

Reply:
Sorry, why outside if? We do not need to update exploration ids in case they are removed from collection progress model. The reason for removal from collection progress model is that they are not part of the collection but are still completed by the user, so they should be present in completed activities model. I have moved this part above the updation of exploration ids in collection progress model, so that removal does not impact completed activities model.

Basically the explorations present in collection progress model are completed by user, so they should be present in completed activities model.

If completed activities model is None, I regenerate it.

If the deleted flag is true, I just revert the flag.

Reply:
Ah, I think moving this upwards answers my question. Thanks!